### PR TITLE
wasm-executor: Support growing the memory

### PR DESCRIPTION
**Pull request by bkchr**
This pull request adds support for growing the wasm memory while executing a runtime function. The pull request doesn't change the old default behavior which was to have a fixed amount of heap pages. By default will still only allocate a fixed amount of memory in total for the wasm memory. The new "allocation mode" that is being added is called "dynamic". If this is enabled, the allocator is allowed to grow the wasm memory up to the upper maximum of what wasm32 supports (4GB). Besides these changes there is a new `CallContext` that needs to passed to the wasm executor call function. Based on this context (onchain/offchain) the executor can use a different heap pages limit as configured on startup. To make it easy to configure the `WasmExecutor` a new `WasmExecutorBuilder` is added to simplify the configuration.

# Why supporting for different maximum heap pages, based on the context?

I think that on chain logic will always need to stay in certain limits. This should be done to ensure that we don't run into situations where the we allocate huge amounts of data that we want to read in some offchain context. In this offchain context we then fail because all this data combined doesn't fit into the memory. Another example are Parachains, they have a hard cap of 1GB of memory for the validation phase. So, they need to ensure to stay in this limit while building the PoV. However, in the offchain context we can be more free and open up the possibility to allocate up to 4GB, if configured.

# Enabling dynamic memory

To enable dynamic memory growing a user would need to use the newly introduced `WasmExecutorBuilder` and set the appropriate values via `with_offchain_heap_alloc_strategy`/`with_onchain_heap_alloc_strategy`. By default nothing changes to the old behavior and everything uses the same memory limits as before. It is also advised to not use this right now, especially not for building/importing blocks. Building/importing blocks should always be done in some fixed limits (especially for Parachains which have only a fixed memory size when validating their blocks) to ensure that everyone can import the blocks.

Works towards: https://github.com/paritytech/substrate/issues/10579

polkadot companion: https://github.com/paritytech/polkadot/pull/6730

### CODE REVIEW

File: client/executor/common/src/runtime_blob/runtime_blob.rs [hunk 1]
```diff
@@ -179,8 +171,13 @@ impl RuntimeBlob {
...
-			let max = memory_ty.limits().maximum().map(|max| std::cmp::max(min, max));
+			let initial = memory_ty.limits().initial();
+			let (min, max) = match heap_pages {
+				HeapPages::Dynamic => (initial, None),
+				HeapPages::Max(max) => (initial, Some(max as _)),
+				HeapPages::ExtraMax(extra) =>
+					(initial + extra as u32, Some(initial + extra as u32)),
+			};
 			*memory_ty = MemoryType::new(min, max);
 		}
```
Comment (koute) [inline client/executor/common/src/runtime_blob/runtime_blob.rs:182]: 1. We're ignoring and overriding the innate limit set inside of the `.wasm` blob here. Compatibility-wise I think this probably shouldn't be a problem since we never supported any kind of memory growth and none of our runtimes should have the maximum set (IIRC you do need to explicitly pass a flag to the linker to have it set). That said, maybe it'd be worthwhile to keep this door open for the future and support the limit anyway (but allow it to be overridden in emergencies) or explicitly *reject* any runtime which has is set? I'm not entirely sure myself. (cc @pepyakin)
2. Do we need the `Max` variant? As far as I can see it's currently used only in tests.
3. I think a better name for `Dynamic` would maybe be `DynamicUnlimited`. Or better yet, just merge the `Max` variant into `Dynamic` and add a `maximum_pages: Option<u32>` field.
4. We should return an error if the maximum number of pages is less than `initial`.
5. I think a better name for `ExtraMax` would maybe be `Static` and have the extra pages as a named argument?

File: client/executor/common/src/runtime_blob/runtime_blob.rs [hunk 2]
```diff
@@ -157,18 +157,10 @@ impl RuntimeBlob {
 		Ok(())
 	}
 
-	/// Increases the number of memory pages requested by the WASM blob by
-	/// the given amount of `extra_heap_pages`.
-	///
-	/// Will return an error in case there is no memory section present,
-	/// or if the memory section is empty.
```
Comment (koute) [inline client/executor/common/src/runtime_blob/runtime_blob.rs:164]: This part about when it returns an error you could probably have left. (:

File: client/executor/wasmtime/src/runtime.rs [hunk 3]
```diff
@@ -354,25 +344,20 @@ fn common_config(semantics: &Semantics) -> std::result::Result<wasmtime::Config,
...
-			if max_memory_size % WASM_PAGE_SIZE != 0 {
-				pages += 1;
-			}
-
-			std::cmp::min(MAX_WASM_PAGES, pages)
-		} else {
-			MAX_WASM_PAGES
+		let memory_pages = match semantics.heap_pages {
+			HeapPages::Max(max) => max as u64,
+			_ => MAX_WASM_PAGES,
```
Comment (koute) [inline client/executor/wasmtime/src/runtime.rs:360]: I don't think we should use a wildcard match here. (:

I think it'd be best to just move the logic from these matches into methods on `impl HeapPages`.

**Review by koute — COMMENTED** [review]
Thanks for the PR!

I've done a very quick look through; I'll do a more thorough review later.

File: client/executor/wasmtime/src/host.rs [hunk 4]
```diff
@@ -68,7 +68,7 @@ impl HostState {
 	}
 
 	pub(crate) fn allocation_stats(&self) -> AllocationStats {
-		self.allocator.stats()
+		self.allocator.as_ref().unwrap().stats()
```
Comment (pepyakin) [inline client/executor/wasmtime/src/host.rs:71]: Nit: a brief comment or rustdoc about the unwrap or turn it into an `expect`?

File: client/executor/wasmtime/src/instance_wrapper.rs [hunk 5]
```diff
@@ -123,66 +122,62 @@ impl EntryPoint {
...
-		Store::new(engine, StoreData { limits, host_state: None, memory: None, table: None });
-	if max_memory_size.is_some() {
-		store.limiter(|s| &mut s.limits);
+	fn max_pages(&self) -> Option<u32> {
+		self.0.ty(&self.1).maximum().map(|p| p as _)
 	}
-	store
+}
+
+pub(crate) fn create_store(engine: &wasmtime::Engine) -> Store {
```
Comment (pepyakin) [inline client/executor/wasmtime/src/instance_wrapper.rs:160]: Nit: consider inlining the function?

File: client/executor/wasmtime/src/host.rs [hunk 6]
```diff
@@ -46,7 +46,7 @@ unsafe impl Send for SandboxStore {}
 /// many different host calls that must share state.
 pub struct HostState {
 	sandbox_store: SandboxStore,
-	allocator: FreeingBumpHeapAllocator,
+	allocator: Option<FreeingBumpHeapAllocator>,
```
Comment (pepyakin) [inline client/executor/wasmtime/src/host.rs:49]: Nit: does it make sense to briefly mention the rationale for using the `Option` here?

File: client/executor/src/wasm_runtime.rs [hunk 7]
```diff
@@ -221,7 +221,7 @@ impl RuntimeCache {
 		runtime_code: &'c RuntimeCode<'c>,
 		ext: &mut dyn Externalities,
 		wasm_method: WasmExecutionMethod,
-		default_heap_pages: u64,
+		heap_pages: HeapPages,
```
Comment (pepyakin) [inline client/executor/src/wasm_runtime.rs:224]: IMO updating the rustdoc in order.

File: client/executor/src/native_executor.rs [hunk 8]
```diff
@@ -85,13 +87,123 @@ pub trait NativeExecutionDispatch: Send + Sync {
...
 /// An abstraction over Wasm code executor. Supports selecting execution backend and
 /// manages runtime cache.
 pub struct WasmExecutor<H> {
 	/// Method used to execute fallback Wasm code.
 	method: WasmExecutionMethod,
-	/// The number of 64KB pages to allocate for Wasm execution.
-	default_heap_pages: u64,
+	/// The number of 64KB pages to allocate for Wasm execution for onchain calls.
+	default_onchain_heap_pages: HeapPages,
+	/// The number of 64KB pages to allocate for Wasm execution for offchain calls.
```
Comment (pepyakin) [inline client/executor/src/native_executor.rs:205]: Nit: not sure if it's fair to say that it's a number

**Review by pepyakin — COMMENTED** [review]
Also a brief review.

UPD: nvm, I am being dumb

File: client/executor/common/src/util.rs [hunk 9]
```diff
@@ -50,3 +46,58 @@ pub trait MemoryTransfer {
...
+///
+/// It needs to be allocated using `mmap` otherwise unmapping fails.
+///
+/// Returns `true` when unmapping was successfull.
+pub fn unmap_memory(memory: &[u8]) -> bool {
+	cfg_if::cfg_if! {
+		if #[cfg(target_os = "linux")] {
+			use std::sync::Once;
+
+			unsafe {
```
[the remark this answers is not in the document]
Comment (pepyakin) [reply client/executor/common/src/util.rs:60]: Just to be clear, accepting a raw pointer still warrants indicating the right safety invariants.

File: client/allocator/src/freeing_bump.rs [hunk 10]
```diff
@@ -800,27 +888,33 @@ mod tests {
...
+			.for_each(|ptr| heap.deallocate(&mut mem, ptr).expect("Deallocate 32 byte"));
+
+		assert_eq!(heap.stats.bytes_allocated, PAGE_SIZE - 16);
+		assert_eq!(heap.bumper, PAGE_SIZE - 16);
+
+		// Allocate another 8 byte to use the full heap.
+		heap.allocate(&mut mem, 8).expect("Allocate 8 byte");
 
 		// when
 		// the `bumper` value is equal to `size` here and any
```
Comment (koute) [inline client/allocator/src/freeing_bump.rs:913]: While we're at it we could add an `assert_eq!` here which verifies in code that this comment is true (;

File: client/allocator/src/freeing_bump.rs [hunk 11]
```diff
@@ -913,24 +1007,24 @@ mod tests {
 
 	#[test]
 	fn deallocate_needs_to_maintain_linked_list() {
-		let mut mem = [0u8; 8 * 2 * 4 + ALIGNMENT as usize];
```
Comment (koute) [inline client/allocator/src/freeing_bump.rs:916]: Hmm.... since `Memory` (which is the public interface) can only return `pages()`, and we only support growing the memory in pages these tests are kind-of nonsense now. You create a `MemoryInstance` here with a size of 72, but when `pages()` is called it will return `1` and when `size()` is called it will return `65536`, neither of which is technically true, and should never actually happen in production code. Functionally for the tests I guess it's fine, but from a conceptual and documentation standpoint it doesn't really make much sense. Would be nice to enforce that with appropriate `assert!`s and adjust the tests to reflect that reality. (And maybe change `with_size` to `with_pages`?)

File: client/allocator/src/freeing_bump.rs [hunk 12]
```diff
@@ -601,21 +642,65 @@ impl<'a> Drop for PoisonBomb<'a> {
...
 
+	struct MemoryInstance {
+		data: Vec<u8>,
+		max_wasm_pages: u32,
+	}
+
+	impl MemoryInstance {
+		fn new() -> Self {
+			Self { data: vec![0; PAGE_SIZE as usize], max_wasm_pages: MAX_WASM_PAGES }
+		}
```
Comment (koute) [inline client/allocator/src/freeing_bump.rs:658]: I'd probably remove this and just use `with_size` everywhere in tests to be explicit. (When reading the code I first assumed that `new` will return a zero-sized memory since literally every Rust data structure works like that [except VecDeque])

File: client/allocator/src/lib.rs [hunk 13]
```diff
@@ -27,3 +27,19 @@ mod freeing_bump;
...
+	/// Run the given closure `run` and grant it read access to the raw memory.
+	fn with_access<R>(&self, run: impl FnOnce(&[u8]) -> R) -> R;
+	/// Grow the memory by `additional` pages.
+	fn grow(&mut self, additional: u32) -> Result<(), ()>;
+	/// Returns the current number of pages this memory has allocated.
+	fn pages(&self) -> u32;
+	/// Returns the maximum number of pages this memory is allowed to allocate.
+	///
+	/// If `None` is returned, there is no maximum (besides the maximum defined in the wasm spec).
+	fn max_pages(&self) -> Option<u32>;
```
Comment (koute) [inline client/allocator/src/lib.rs:44]: It would be nice to explicitly spell out in these doc comments

1) how big a page is,
2) what the maximum `max_pages` actually is,
3) and that `max_pages` must be lower or equal to the maximum `max_pages`

**Review by koute — COMMENTED** [review]
Okay, I went through the changes in the allocator. (:

File: client/executor/wasmtime/src/tests.rs [hunk 14]
```diff
@@ -345,14 +346,14 @@ fn test_max_memory_pages(
 	precompile_runtime: bool,
 ) {
```
Comment (koute) [inline client/executor/wasmtime/src/tests.rs:347]: Now that we've merged `max_memory_size` and `extra_heap_pages` this test is kinda confusing and would warrant a cleanup and an expansion adding more cases.

We essentially check three categories of things (based on the unique wat code blobs used here):
1) whether the instantiation succeeds with given `heap_pages` + memory section
2) whether growing the memory succeeds with given `heap_pages` + memory section
3) whether growing the memory fails with given `heap_pages` + memory_section

So could you maybe add helper function for each "category" of tests (each would wrap a whole `try_instantiate` call) and rewrite the test in this manner? Something like this:

```
let assert_instantiation_ok = |heap_pages: HeapPages, min: u32 max: Option<u32>| {
    // ...
};
let assert_grow_ok = ...;
let assert_grow_fails = ...;
// ...

assert_instantiation_ok(HeapPages::ExtraMax(1024), 64511, None);
assert_instantiation_fails(HeapPages::ExtraMax(1025), 64511, None);
assert_grow_ok(HeapPages::ExtraMax(1024 + 24), 0, None);
// ...
```

Basically so that we can easily cover every combination of `HeapPages` in every boundary scenario.

File: client/service/test/src/client/mod.rs [hunk 15]
```diff
@@ -113,6 +113,7 @@ fn construct_block(
 		Default::default(),
 		&runtime_code,
 		task_executor.clone() as Box<_>,
+		CallContext::Offchain,
```
Comment (koute) [inline client/service/test/src/client/mod.rs:116]: Shouldn't this be `Onchain` since this is supposed to be block construction?

**Comment by kianenigma** [comment]
Once this is done, I have a great application for it on westend: we increase nominators, and we let the offchain workers do the NpoS computation. Currently, it will break very soon, or even with the current nominator count, running out of memory. We expect this to now work.

**Comment by koute** [comment]
not stale

**Comment by koute** [comment]
not stale

**Comment by kianenigma** [comment]
> Once this is done, I have a great application for it on westend: we increase nominators, and we let the offchain workers do the NpoS computation. Currently, it will break very soon, or even with the current nominator count, running out of memory. We expect this to now work.

What's the status of this?

**Comment by bkchr** [comment]
> > Once this is done, I have a great application for it on westend: we increase nominators, and we let the offchain workers do the NpoS computation. Currently, it will break very soon, or even with the current nominator count, running out of memory. We expect this to now work.
> 
> What's the status of this?

Yeah I will resurrect it in the next days :see_no_evil:

File: client/service/test/src/client/mod.rs (hunk 15 above)
Comment (bkchr) [reply client/service/test/src/client/mod.rs:116]: I mean it is just some test. :D Not thatttt important :P

File: client/allocator/src/freeing_bump.rs (hunk 11 above)
Comment (bkchr) [reply client/allocator/src/freeing_bump.rs:916]: We are not testing here the public interface of `Memory`. We are testing here some specific internals that should ensure that we free up space accordingly. While you are right that this will not happen in reality (which was already true before this pr), we are just testing here some specific part of the implementation. I think we should keep it.

File: client/allocator/src/freeing_bump.rs [hunk 16]
```diff
@@ -538,35 +572,42 @@ impl FreeingBumpHeapAllocator {
...
+		})
 	}
-	fn size(&self) -> u32 {
-		u32::try_from(self.len()).expect("size of Wasm linear memory is <2^32; qed")
+
+	fn size(&self) -> u64 {
+		let len = self.pages() as u64 * PAGE_SIZE as u64;
+		u64::try_from(len).expect("size of Wasm linear memory is <=2^32; qed")
 	}
 }
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/allocator/src/freeing_bump.rs:612]: > 2\. The `try_from` here is useless as you're converting an `u64` into an `u64` here. (:

:facepalm: :see_no_evil: 



> 1. I'd probably keep this returning `u32` (since it should never physically return anything larger than `u32` so it's a little confusing to have it return `u64`)

2^32 doesn't fit into u32. I mean we can address the entire memory, but to represent its size this is not possible. Or my brain here is doing some stupid thing?

File: client/allocator/src/freeing_bump.rs [hunk 17]
```diff
@@ -503,25 +509,53 @@ impl FreeingBumpHeapAllocator {
...
+					log::debug!(
+						target: LOG_TARGET,
+						"Number of required pages({required_pages}) is greater \
+						 than the maximum number of pages({max_pages}).",
+					);
+					return Err(Error::AllocatorOutOfSpace)
+				} else if memory.grow(next_pages - pages).is_err() {
+					log::error!(
+						target: LOG_TARGET,
+						"Failed to grow memory from {pages} pages to {next_pages} pages",
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/allocator/src/freeing_bump.rs:543]: Not sure why. This number is not important here? I mean I would then also need to print `min_grow`, but in the end nothing of that is important IMO.

[the remark this answers is not in the document]
Comment (bkchr) [reply client/allocator/src/freeing_bump.rs:543]: The error is just `()`

File: client/allocator/src/freeing_bump.rs [hunk 18]
```diff
@@ -503,25 +509,53 @@ impl FreeingBumpHeapAllocator {
...
+
+				return Err(Error::AllocatorOutOfSpace)
+			} else {
+				// Let us growth by at least pages * 2, but ensure we stay in the allowed maximum
+				// number of pages.
+				let min_grow = min(pages * 2, max_pages);
+				let next_pages = max(min_grow, required_pages);
+
+				if required_pages > max_pages {
+					log::debug!(
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/allocator/src/freeing_bump.rs:534]: Because that just means that the runtime tries to allocate too much memory, but this isn't really any application error on the node side. We just prevent this.

File: client/executor/common/src/runtime_blob/runtime_blob.rs (hunk 1 above)
Comment (bkchr) [reply client/executor/common/src/runtime_blob/runtime_blob.rs:182]: > 2\. Do we need the `Max` variant? As far as I can see it's currently used only in tests.

For PVF execution where we want some fixed maximum.

> 1. We're ignoring and overriding the innate limit set inside of the `.wasm` blob here. Compatibility-wise I think this probably shouldn't be a problem since we never supported any kind of memory growth and none of our runtimes should have the maximum set

Let me throw an error if this is set.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2391740

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2391738

File: client/executor/src/integration_tests/linux.rs [hunk 19]
```diff
@@ -18,33 +18,64 @@
...
 use self::smaps::Smaps;
 
 #[test]
+fn memory_consumption_interpreted() {
+	let _ = sp_tracing::try_init_simple();
+
+	if std::env::var("RUN_TEST").is_ok() {
+		memory_consumption(WasmExecutionMethod::Interpreted);
+	} else {
+		// We need to run the test in isolation, to not getting interfered by the other tests.
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/executor/src/integration_tests/linux.rs:37]: Tests are run in parallel and then they can interfere each other AFAIR.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2396587

File: client/executor/wasmtime/src/tests.rs (hunk 14 above)
Comment (bkchr) [reply client/executor/wasmtime/src/tests.rs:347]: I didn't fully get what you wanted. But I have seen what you meant on what doesn't make any sense. I removed the nonsense tests. Testing instantion in general doesn't make that much sense as we overwrite the memory definition.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2399068

File: client/allocator/src/freeing_bump.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (koute) [reply client/allocator/src/freeing_bump.rs]: > 2^32 doesn't fit into u32. I mean we can address the entire memory, but to represent its size this is not possible. Or my brain here is doing some stupid thing?

Hmm... I guess this is a good point. Yeah, you're right.

File: client/executor/src/integration_tests/linux.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (koute) [reply client/executor/src/integration_tests/linux.rs:37]: > Tests are run in parallel and then they can interfere each other AFAIR.

Yes but the test is fetching the RSS of a *specific* map which is unique to that test, this part specifically:

```rust
fn probe_rss(instance: &dyn sc_executor_common::wasm_runtime::WasmInstance) -> usize {
    let base_addr = instance.linear_memory_base_ptr().unwrap() as usize;
    Smaps::new().get_rss(base_addr).expect("failed to get rss")
}
```

It gets the base address of the linear memory (which should be unique to that instance and test), it grabs smaps, and then it finds the specific map which corresponds to that linear memory block and returns the RSS *only* for that map, and not for the whole process. So it should be fine to run this within the same process (unless there's a bug).

File: client/executor/runtime-test/src/lib.rs [hunk 20]
```diff
@@ -344,6 +344,15 @@ sp_core::wasm_export_functions! {
...
 
+	fn allocate_two_gigabyte() -> u32 {
+		let mut data = Vec::new();
+		for _ in 0..205 {
+			data.push(Vec::<u8>::with_capacity(10 * 1024 * 1024));
+		}
+
+		data.iter().map(|d| d.capacity() as u32).sum()
+	}
+
```
[the remark this answers is not in the document]
Comment (koute) [reply client/executor/runtime-test/src/lib.rs:339]: Hmm.... okay, you're right. I don't know why I thought this allocates less; must have had a late night brainfart or something and messed up the math. Sorry.

File: client/allocator/src/freeing_bump.rs [hunk 21]
```diff
@@ -503,25 +518,53 @@ impl FreeingBumpHeapAllocator {
...
+
+				return Err(Error::AllocatorOutOfSpace)
+			}
+
+			// Let us grow by at least pages * 2, but ensure we stay within the allowed maximum
+			// number of pages.
+			let min_grow = min(pages * 2, max_pages);
+			let next_pages = max(min_grow, required_pages);
+
+			if required_pages > max_pages {
```
Comment (koute) [inline client/allocator/src/freeing_bump.rs:543]: Doesn't really matter, but this `if` could be moved slightly earlier since it's checking `required_pages` and not `next_pages`.

**Review by koute — COMMENTED** [review]
Looks mostly good, just some minor stuff!

File: client/executor/runtime-test/src/lib.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply client/executor/runtime-test/src/lib.rs:339]: Haha no problem :D

File: primitives/sandbox/Cargo.toml [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+	"codec/std",
+	"log/std",
+	"sp-core/std",
+	"sp-io/std",
+	"sp-std/std",
+	"sp-wasm-interface/std",
+	"wasmi/std",
+]
+strict = []
+wasmer-sandbox = []
```
[the remark this answers is not in the document]
Comment (bkchr) [reply primitives/sandbox/Cargo.toml:40]: Welcome back!

File: client/executor/wasmi/src/lib.rs [hunk 23]
```diff
@@ -522,6 +504,8 @@ pub struct WasmiInstance {
 	instance: ModuleRef,
 	/// The memory instance of used by the wasm module.
 	memory: MemoryRef,
+	/// Is the memory zeroed?
+	memoy_zeroed: bool,
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/executor/wasmi/src/lib.rs:508]: Yes :P

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2407384

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2407385

**Review by koute — APPROVED** [review]
LGTM

File: client/executor/wasmi/src/lib.rs [hunk 24]
```diff
@@ -574,7 +560,12 @@ impl WasmiInstance {
 			self.allow_missing_func_imports,
 			self.missing_functions.clone(),
 			allocation_stats,
-		)
+		);
+
+		// If we couldn't unmap it, erase the memory.
```
Comment (michalkucharczyk) [inline client/executor/wasmi/src/lib.rs:565]: I don't understand this comment. And dq: what is the point of having memory_zeored?

File: test-utils/runtime/src/lib.rs [hunk 25]
```diff
@@ -1358,7 +1358,12 @@ mod tests {
 
 		// Try to allocate 1024k of memory on heap. This is going to fail since it is twice larger
 		// than the heap.
-		let ret = client.runtime_api().vec_with_capacity(&block_id, 1048576);
+		let ret = client.runtime_api().vec_with_capacity_with_context(
+			&block_id,
```
Comment (michalkucharczyk) [inline test-utils/runtime/src/lib.rs:1362]: Will have a conflict here:
```suggestion
			client.chain_info().best_hash,
```

**Comment by michalkucharczyk** [comment]
DQ: Having different memory model for benchmarking (`dynamic`) and for runtime execution (`static`) may lead to wrongly estimated cost of memory. For example in case of allocation failure the runtime code may fallback to more expensive algorithm (imagine doing something in place instead of using extra buffer). Not sure if this is applicable here.

File: client/executor/wasmi/src/lib.rs (hunk 24 above)
Comment (davxy) [reply client/executor/wasmi/src/lib.rs:565]: I leave the answer to @bkchr but probably is to erase the program bss? I.e. global varables are assumed to have zero value on program entry.

Comment (davxy) [reply client/executor/wasmi/src/lib.rs:565]: (For the next program that will run)

File: client/executor/wasmi/src/lib.rs [hunk 26]
```diff
@@ -549,14 +533,16 @@ impl WasmiInstance {
...
-		})?;
+		if !self.memory_zeroed {
+			// First, zero initialize the linear memory.
+			self.memory.erase().map_err(|e| {
+				// Snapshot restoration failed. This is pretty unexpected since this can happen
+				// if some invariant is broken or if the system is under extreme memory pressure
+				// (so erasing fails).
+				error!(target: "wasm-executor", "snapshot restoration failed: {}", e);
+				WasmError::ErasingFailed(e.to_string())
+			})?;
```
Comment (michalkucharczyk) [inline client/executor/wasmi/src/lib.rs:544]: should we set the `memory_zeroed` afterwards?

File: client/executor/common/src/runtime_blob/runtime_blob.rs [hunk 27]
```diff
@@ -179,8 +174,17 @@ impl RuntimeBlob {
 			return Err(WasmError::Other("memory section is empty".into()))
 		}
 		for memory_ty in memory_section.entries_mut() {
-			let min = memory_ty.limits().initial().saturating_add(extra_heap_pages);
-			let max = memory_ty.limits().maximum().map(|max| std::cmp::max(min, max));
+			let initial = memory_ty.limits().initial();
+			let (min, max) = match heap_alloc_strategy {
+				HeapAllocStrategy::Dynamic { maximum_pages } => {
+					// Ensure `initial <= maximum_pages`
+					(maximum_pages.map(|m| m.min(initial)).unwrap_or(initial), maximum_pages)
```
[the remark this answers is not in the document]
Comment (michalkucharczyk) [reply client/executor/common/src/runtime_blob/runtime_blob.rs:181]: Isn't it intentional? We don't want `initial` to be greater then `max`.

[the remark this answers is not in the document]
Comment (davxy) [reply client/executor/common/src/runtime_blob/runtime_blob.rs:181]: Indeed. But what I was saying is that if the program requires an initial greater than `maximum_pages` we end up giving to the program less memory than how much it requires. I may have missed something anyway 🤔 And that maybe is intentional

**Review by michalkucharczyk — APPROVED** [review]

**Review by skunert — APPROVED** [review]

File: client/executor/wasmi/src/lib.rs (hunk 24 above)
Comment (koute) [reply client/executor/wasmi/src/lib.rs:565]: Every call into the runtime needs to start with cleared memory so that one invocation can't accidentally depend on the state from the previous invocation. (We didn't always do it and in fact we used not to clear the memory, but this lead to horrible bugs [which I had to investigate :P].)

**Comment by bkchr** [comment]
> DQ: Having different memory model for benchmarking (`dynamic`) and for runtime execution (`static`) may lead to wrongly estimated cost of memory. For example in case of allocation failure the runtime code may fallback to more expensive algorithm (imagine doing something in place instead of using extra buffer). Not sure if this is applicable here.

When the allocation fails, we stop the execution. The runtime doesn't see that the allocation failed.

File: client/allocator/src/lib.rs [hunk 28]
```diff
@@ -27,3 +27,25 @@ mod freeing_bump;
...
+///
+/// Memory of wasm is allocated in pages. A page has a constant size of 64KiB. The maximum allowed
+/// memory size as defined in the wasm specification is 4GiB (65536 pages).
+pub trait Memory {
+	/// Run the given closure `run` and grant it write access to the raw memory.
+	fn with_access_mut<R>(&mut self, run: impl FnOnce(&mut [u8]) -> R) -> R;
+	/// Run the given closure `run` and grant it read access to the raw memory.
+	fn with_access<R>(&self, run: impl FnOnce(&[u8]) -> R) -> R;
+	/// Grow the memory by `additional` pages.
+	fn grow(&mut self, additional: u32) -> Result<(), ()>;
```
[the remark this answers is not in the document]
Comment (koute) [reply client/allocator/src/lib.rs:41]: If you use `Result` it's immediately obvious that the function can error out, and that `Result::Err` is an error. And you'll get a warning if the `Result` is not handled (since it's marked with `#[must_use]`). With `bool` there's no such convention.

And of course with `Result` you can use e.g. `map_err` and `?`.

[the remark this answers is not in the document]
Comment (bkchr) [reply client/allocator/src/lib.rs:41]: Good question, kind of  taste? I mean it is a failure and a simple `bool` feels too much like C, but semantically it doesn't make any difference.

File: client/allocator/src/freeing_bump.rs [hunk 29]
```diff
@@ -601,21 +657,72 @@ impl<'a> Drop for PoisonBomb<'a> {
...
+		assert_eq!(pages_from_size(2 * 65536).unwrap(), 2);
+		assert_eq!(pages_from_size(2 * 65536 + 1).unwrap(), 3);
+	}
+
 	#[test]
 	fn should_allocate_properly() {
 		// given
-		let mut mem = [0u8; PAGE_SIZE as usize];
+		let mut mem = MemoryInstance::with_pages(1);
 		let mut heap = FreeingBumpHeapAllocator::new(0);
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/allocator/src/freeing_bump.rs:722]: Yeah should probably now be achievable, before with just operating on the bytes it would have required a trait (which we now got). Maybe some possible future refactor, but is also not such a huge problem.

File: client/executor/common/src/runtime_blob/runtime_blob.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply client/executor/common/src/runtime_blob/runtime_blob.rs:181]: Yes that is intentional. If something requires more, then it will probably fail at execution.

File: client/allocator/src/freeing_bump.rs (hunk 29 above)
[the remark this answers is not in the document]
Comment (koute) [reply client/allocator/src/freeing_bump.rs:722]: In case of `wasmtime` the actual memory is owned by the executor, so to access it you also need a mutable handle to the context, but you're also storing the `FreeingBumpHeapAllocator` inside the context, and you can't have two mutable references to the same thing. That's why Basti has to store the allocator inside of an `Option` and temporarily `take()` it in `client/executor/wasmtime/src/host.rs`.

Technically you *could* add a field inside of the `FreeingBumpHeapAllocator` which would store a handle to the memory, but it still wouldn't own the memory (because it's owned by `wasmtime` and the `Memory` it exposes is just essentially an `usize` index into `wasmtime`'s internal store) and you still would need `wasmtime`'s context to access it. So there isn't really much point in doing that.

**Comment by bkchr** [comment]
bot merge
